Add io.frappe.books

kskarthik opened the pull request:
<!-- ⚠️  The submission PR must be against the `new-pr` branch ⚠️  -->

### Please confirm your submission meets all the criteria

<!-- Please replace each `[ ]` with `[X]` when the step is complete -->

- [x] Please describe the application briefly.
  Free Desktop book-keeping software for small businesses and freelancers.

- [x] The domain used for the application ID is [controlled by the application developers][appid-domain] and the [application id guidelines][appid] are followed.
- [x] I have read the and followed all the [Submission and licence requirements][reqs].
- [x] I have [built][build] and tested the submission locally.
- [ ] I am an author/developer/upstream contributor of the project. If not, I contacted upstream developers about this submission. **Link:** https://github.com/frappe/books/issues/206#issuecomment-2410892420


<!-- ⚠️  Don't modify anything below this line ⚠️  -->

[appid-domain]: https://docs.flathub.org/docs/for-app-authors/requirements/#control-over-domain-or-repository
[appid]: https://docs.flathub.org/docs/for-app-authors/requirements#application-id
[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission

hfiguiere commented on io.frappe.books.yml:
```diff
@@ -0,0 +1,46 @@
+      - type: archive
+        only-arches:
+          - x86_64
+        url: "https://github.com/frappe/books/releases/download/v0.23.0/frappe-books-0.23.0.x86_64.rpm"
```
why don't you build it from source?

kskarthik replied:
I have tried. But npm install fails with an error (after generating the sources list from flatpak-builder-tools)

kskarthik replied:
Many thanks for the insighful comments. I have learnt a lot more from this PR

This is my first flatpak build & flathub submissions and i greatly appreciate your guidance :)

hfiguiere commented on io.frappe.books.yml:
```diff
@@ -0,0 +1,46 @@
+    buildsystem: simple
+    build-commands:
+      # create an app dir in the flatpak's bin
+      - mkdir -p /app/bin/books
```
the app dir should not be in the `bin` directory.

hfiguiere commented on io.frappe.books.yml:
```diff
@@ -0,0 +1,46 @@
+      # create an app dir in the flatpak's bin
+      - mkdir -p /app/bin/books
+      # move the contents from the builddir to flatpak's filesystem
+      - mv /run/build/books/'Frappe Books'/* /app/bin/books/
+      # Install app wrapper
+      - install -Dm755 /run/build/books/run.sh /app/bin/
+      # Install desktop icon
+      - install -Dm644 /run/build/books/share/icons/hicolor/256x256/apps/frappe-books.png /app/share/icons/hicolor/256x256/apps/${FLATPAK_ID}.png
```
```suggestion
      - mv ./'Frappe Books'/* /app/bin/books/
      # Install app wrapper
      - install -Dm755 ./run.sh /app/bin/
      # Install desktop icon
      - install -Dm644 ./share/icons/hicolor/256x256/apps/frappe-books.png /app/share/icons/hicolor/256x256/apps/${FLATPAK_ID}.png
```

hfiguiere commented on io.frappe.books.yml:
```diff
@@ -0,0 +1,46 @@
+  - --socket=x11
+  - --socket=pulseaudio
+  # - --share=network
+  # only access the app's dir on the host
+  - --filesystem=~/Frappe Books:create
```
no. It should use `--persist` instead.

hfiguiere commented on io.frappe.books.desktop:
you mean the RPM come without one?

hfiguiere commented on io.frappe.books.desktop:
```diff
@@ -0,0 +1,10 @@
+Name=Books
+Comment= Open source Accounting Software
+Icon=io.frappe.books
+Exec=/app/bin/run.sh
```
it is NEVER an absolute path.

hfiguiere commented on flathub.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "only-arches": ["x86_64"]
```
this is disappointing.

ninjadev64 replied:
Why don't you be a bit nicer @hfiguiere?

bbhtt commented:
can you move appdata and desktop file to https://github.com/frappe/ ?

kskarthik commented:
> can you move appdata and desktop file to https://github.com/frappe/ ?

you mean this repo into the frappe's github project ? I do not have permissions

bbhtt commented:
Then open a PR adding them.

kskarthik commented:
@bbhtt from what i understood, there should be a repo at `github.com/frappe/flathub` with these files. am i correct ?

bbhtt commented:
> with these files

these are supposed to be from upstream, not supposed to be in the repo created at flathub

kskarthik commented:
you mean adding these flatpak build files to the https://github.com/frappe/books repo via a merge request?

bbhtt commented:
Not all of them, just the metainfo and desktop file.

kskarthik commented:
> Not all of them, just the metainfo and desktop file.

Done https://github.com/frappe/books/pull/980

hfiguiere commented:
> > Not all of them, just the metainfo and desktop file.
> 
> Done [frappe/books#980](https://github.com/frappe/books/pull/980)

The manifest isn't necessary. And neither of the other two files are flatpak specific so their location isn't correct.

kskarthik commented:
> > > Not all of them, just the metainfo and desktop file.
> > 
> > 
> > Done [frappe/books#980](https://github.com/frappe/books/pull/980)
> 
> The manifest isn't necessary. And neither of the other two files are flatpak specific so their location isn't correct.

@hfiguiere The correct location should be in the root of the project under `flatpak/` ?

And required file is only the *.appdata.xml  ?

hfiguiere commented:
> @hfiguiere The correct location should be in the root of the project under `flatpak/` ?

Since I said that it's not flatpak specific, why would a directory named flatpak be a good place. For the rest it's up to upstream, but don't steer them in the wrong direction,

 
> And required file is only the *.appdata.xml ?

The .desktop too, and if they don't ship one in their package it's a real bug regardless of flatpak.

kskarthik commented:
> The .desktop too, and if they don't ship one in their package

There is no desktop file in the upstream repo.

hfiguiere commented on io.frappe.books.desktop:
This file should be removed and the one from the package used.

`/usr/share/applications/frappe-books.desktop`

kskarthik replied:
inside the flatpak, you mean?

hfiguiere replied:
this is from the RPM....

kskarthik commented:
> Since I said that it's not flatpak specific, why would a directory named flatpak be a good place. For the rest it's up to upstream, but don't steer them in the wrong direction,


I have changed the paths & only placed the metainfo & desktop files as suggested in one of the later comments https://github.com/frappe/books/pull/980/files

bbhtt commented on io.frappe.books.yml:
```diff
@@ -0,0 +1,39 @@
+    buildsystem: simple
+    build-commands:
+      # copy the contents from the builddir to the flatpak's filesystem
+      - cp -r ./'Frappe Books'/* /app/bin/
```
```suggestion
      - cp -ra opt/'Frappe Books'/ /app
```

bbhtt commented on io.frappe.books.yml:
```diff
@@ -0,0 +1,39 @@
+      # copy the contents from the builddir to the flatpak's filesystem
+      - cp -r ./'Frappe Books'/* /app/bin/
+      # Install app wrapper
+      - install -Dm755 ./run.sh /app/bin/
```
```suggestion
      - install -Dm755 run.sh -t /app/bin/
```

bbhtt commented on io.frappe.books.yml:
```diff
@@ -0,0 +1,39 @@
+      # Install app wrapper
+      - install -Dm755 ./run.sh /app/bin/
+      # Install desktop icon
+      - install -Dm644 ./share/icons/hicolor/256x256/apps/frappe-books.png /app/share/icons/hicolor/256x256/apps/${FLATPAK_ID}.png
+      # install the desktop file
+      - install -Dm644 ${FLATPAK_ID}.desktop /app/share/applications/${FLATPAK_ID}.desktop
```
```suggestion
      - cp -r usr/share /app
      - desktop-file-edit --set-key="Exec" --set-value="run.sh %U" ${FLATPAK_DEST}/share/applications/${FLATPAK_ID}.desktop
```

bbhtt commented on io.frappe.books.yml:
```diff
@@ -0,0 +1,39 @@
+        dest-filename: run.sh
+        commands:
+          - zypak-wrapper.sh /app/bin/frappe-books "$@"
+      - type: file
+        path: io.frappe.books.desktop
```
```suggestion
```

bbhtt commented on io.frappe.books.yml:
```diff
@@ -0,0 +1,39 @@
+base: org.electronjs.Electron2.BaseApp
+base-version: '24.08'
+command: run.sh
+separate-locales: false
+finish-args:
```
```suggestion
separate-locales: false
rename-desktop-file: frappe-books.desktop
rename-icon: frappe-books
finish-args:
```

bbhtt commented on io.frappe.books.yml:
```diff
@@ -0,0 +1,39 @@
+      - type: script
+        dest-filename: run.sh
+        commands:
+          - zypak-wrapper.sh /app/bin/frappe-books "$@"
```
```suggestion
          - zypak-wrapper.sh /app/Frappe\ Books/frappe-books "$@"
```

bbhtt commented on io.frappe.books.yml:
```diff
@@ -0,0 +1,38 @@
+        url: "https://github.com/frappe/books/releases/download/v0.23.0/frappe-books-0.23.0.x86_64.rpm"
+        sha256: fb3f8b5db499cc892cbaedfb23bbef99fa22bf89017d973eea91d7a554657b68
+
+      - type: script
+        dest-filename: run.sh
+        commands:
+          - zypak-wrapper.sh /app/Frappe\ Books/frappe-books "$@"
```
```suggestion
      - type: script
        dest-filename: run.sh
        commands:
          - zypak-wrapper.sh /app/Frappe\ Books/frappe-books "$@"
      - type: file
        path: io.frappe.books.appdata.xml
```

bbhtt commented on io.frappe.books.yml:
```diff
@@ -0,0 +1,38 @@
+      - install -Dm755 run.sh -t /app/bin/
+      # Install desktop icon
+      - cp -r usr/share /app
+      - desktop-file-edit --set-key="Exec" --set-value="run.sh %U" ${FLATPAK_DEST}/share/applications/${FLATPAK_ID}.desktop
+    sources:
```
```suggestion
      - desktop-file-edit --set-key="Exec" --set-value="run.sh %U" ${FLATPAK_DEST}/share/applications/${FLATPAK_ID}.desktop
      - install -Dm0644 io.frappe.books.appdata.xml -${FLATPAK_DEST}/share/appdata/${FLATPAK_ID}.appdata.xml
    sources:
```

bbhtt commented:
remove the desktop file from here.

bbhtt commented:
bot, build io.frappe.books

kskarthik commented:
@bbhtt the build fails locally with the patches we introduced. I fixed most of them locally but i hit the wall at. please help me with the below err

```
❯ ./build-flatpak.sh 
Dependency Sdk: org.freedesktop.Sdk 24.08
Updating org.freedesktop.Sdk/x86_64/24.08

Nothing to do.
Dependency Runtime: org.freedesktop.Platform 24.08
Updating org.freedesktop.Platform/x86_64/24.08

Nothing to do.
Dependency Base: org.electronjs.Electron2.BaseApp 24.08
Updating org.electronjs.Electron2.BaseApp/x86_64/24.08

Nothing to do.
Emptying app dir 'build'
Downloading sources
Starting build of io.frappe.books
Cache hit for books, skipping build
Cache miss, checking out last cache hit
Cleaning up
Renaming io.frappe.books.appdata.xml to share/metainfo/io.frappe.books.metainfo.xml
Renaming frappe-books.desktop to io.frappe.books.desktop
Renaming icon hicolor/128x128/apps/frappe-books.png to hicolor/128x128/apps/io.frappe.books.png
Renaming icon hicolor/16x16/apps/frappe-books.png to hicolor/16x16/apps/io.frappe.books.png
Renaming icon hicolor/256x256/apps/frappe-books.png to hicolor/256x256/apps/io.frappe.books.png
Renaming icon hicolor/32x32/apps/frappe-books.png to hicolor/32x32/apps/io.frappe.books.png
Renaming icon hicolor/48x48/apps/frappe-books.png to hicolor/48x48/apps/io.frappe.books.png
Renaming icon hicolor/512x512/apps/frappe-books.png to hicolor/512x512/apps/io.frappe.books.png
Renaming icon hicolor/64x64/apps/frappe-books.png to hicolor/64x64/apps/io.frappe.books.png
Rewriting contents of io.frappe.books.desktop
Running appstreamcli compose
AppStream Compose binary '/usr/libexec/appstreamcli-compose' was not found! Can not continue.
You may be able to install the AppStream Compose addon via: `sudo appstreamcli install org.freedesktop.appstream.compose`
Error: ERROR: appstreamcli compose failed: Child process exited with code 4
```

bbhtt commented:
can you push the fixes? That's an issue on your end, probably a missing package or a dependency.

kskarthik commented:
> can you push the fixes? That's an issue on your end, probably a missing package or a dependency.

just pushed now

bbhtt commented:
bot, build io.frappe.books

kskarthik commented:
@bbhtt installing appstearm compose package fixed the issue. But now had a new one

```
Exporting share/icons/hicolor/64x64/apps/io.frappe.books.png
Exporting share/metainfo/io.frappe.books.metainfo.xml
Please review the exported files and the metadata
Committing stage finish to cache
Exporting io.frappe.books to repo
error: /home/sk/my/projects/flathub/build/export/share/icons/hicolor/512x512/apps/io.frappe.books.png is not a valid icon: Imag
e too large (513x513). Max. size 512x512

Export failed: Child process exited with code 1

bbhtt commented:
Yes max is 512 px, you can resize it with `ffmpeg -i source.png -vf scale=512:512 output.png`, then install output instead of that.

kskarthik commented:
> Yes max is 512 px, you can resize it with `ffmpeg -i source.png -vf scale=512:512 output.png`, then install output instead of that.

Fixed it!

bbhtt commented:
If I click Create Demo it's stuck at setting up an instance...

kskarthik commented:
> If I click Create Demo it's stuck at setting up an instance...

it takes a while to populate data

kskarthik commented:
@bbhtt To summarize:

The app needs access to store data (create or load existing company's db) which is satisfied with` - --persist=Frappe Books`

But "create demo" action expects the user to provide a directory to save a demo company's data. If it creates a new demo db in it's standard storage path that could solve the problem

Few ways to mitigate:

1. discuss with upstream to pick the "~/Frappe Books" as the default path for "create demo"

2. Allow the flatpak full access the user's $HOME by default

3. add a note to the user to manually grant permission to the app to access their home via cli or flatseal

bbhtt commented:
You can make it default to `~/Frappe Books` and add `--filesystem=~/Frappe Books:create`

kskarthik commented:
> You can make it default to `~/Frappe Books` and add `--filesystem=~/Frappe Books:create`

I have added the suggested flag to the finish-args, Not sure how to make the file picker choose `~/Frappe Books` when invoked

bbhtt commented:
You have to patch it.

kskarthik commented:
> You have to patch it.

then, i guess i have to build the app from source inside flatpak  to patch it

kskarthik commented:
Just observed: the app allows the user to load db files from any of the directory in home for the options

1. Existing company
2. Create demo

but new company is saved to `~/Frappe Books` & launched by default (no confirmation or notification is given about the path)

bbhtt commented:
Access to a single directory is preferred over entire home, if you can't make it work then add `home`.

bbhtt commented:
bot, build io.frappe.books
